### PR TITLE
WIP: cmd/setup-etcd-environment: add etcd pivot logic

### PR DESCRIPTION
**Pull request by hexfusion**
This PR adds logic required to scale up during install via openshift-etcd-operator. More details soon..

required by: https://github.com/openshift/cluster-etcd-operator/pull/19

### CODE REVIEW

**Comment by cgwalters** [comment]
Would be nice to avoid the term "pivot" for this since it's highly associated here with the `pivot` command used in RHCOS (4.1).

**Comment by hexfusion** [comment]
> Would be nice to avoid the term "pivot" for this since it's highly associated here with the `pivot` command used in RHCOS (4.1).

not an issue

File: cmd/setup-etcd-environment/run.go [hunk 1]
```diff
@@ -86,9 +105,54 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
+			panic(err.Error())
+		}
+		client, err := kubernetes.NewForConfig(clientConfig)
+		if err != nil {
+			return fmt.Errorf("error creating client: %v", err)
+		}
+		duration := 10 * time.Second
+		// wait forever for success and retry every duration interval
+		wait.PollInfinite(duration, func() (bool, error) {
+			result, err := client.CoreV1().ConfigMaps("openshift-etcd").Get("scaling-lock", metav1.GetOptions{})
```
Comment (beekhof) [inline cmd/setup-etcd-environment/run.go]: At least in this usage, "scaling-lock" seems misnamed.
Perhaps "known-members"?

Comment (hexfusion) [reply cmd/setup-etcd-environment/run.go]: yeah, this configmap originally was a distributed lock similar to the leader election process. But I did not find that to be necessary will consider renaming.

File: cmd/setup-etcd-environment/run.go [hunk 2]
```diff
@@ -86,9 +105,54 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
-		"DISCOVERY_SRV": runOpts.discoverySRV,
-	}, out, true); err != nil {
+	export := make(map[string]string)
+	etcdName := os.Getenv("ETCD_NAME")
+	if etcdName == "" {
+		return fmt.Errorf("environment variable ETCD_NAME has no value")
+	}
+	var e EtcdScaling
+	if runOpts.pivot {
+		clientConfig, err := rest.InClusterConfig()
```
Comment (abhinavdahiya) [inline cmd/setup-etcd-environment/run.go]: Does this mean that it requires k8s networking to be up to talk to apiservers... That would be difficult in recovery scenarios

File: cmd/setup-etcd-environment/run.go [hunk 3]
```diff
@@ -86,9 +105,54 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
+		client, err := kubernetes.NewForConfig(clientConfig)
+		if err != nil {
+			return fmt.Errorf("error creating client: %v", err)
+		}
+		duration := 10 * time.Second
+		// wait forever for success and retry every duration interval
+		wait.PollInfinite(duration, func() (bool, error) {
+			result, err := client.CoreV1().ConfigMaps("openshift-etcd").Get("scaling-lock", metav1.GetOptions{})
+			if err != nil {
+				klog.Errorf("error creating client %v", err)
```
Comment (kikisdeliveryservice) [inline cmd/setup-etcd-environment/run.go]: wondering why you used klog here?

File: cmd/setup-etcd-environment/run.go [hunk 4]
```diff
@@ -86,9 +105,54 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
+			}
+			if err := json.Unmarshal([]byte(result.Annotations[EtcdScalingAnnotationKey]), &e); err != nil {
+				klog.Errorf("error decoding result %v", err)
+				return false, nil
+			}
+			if e.Metadata.Name != etcdName {
+				klog.Errorf("could not find self in scaling-lock")
+				return false, nil
+			}
+			members := e.Members
```
Comment (kikisdeliveryservice) [inline cmd/setup-etcd-environment/run.go]: do we need this placeholder var? would using e.Members directly cause an issue?

File: cmd/setup-etcd-environment/run.go [hunk 5]
```diff
@@ -86,9 +105,54 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
 
-	if err := writeEnvironmentFile(map[string]string{
-		"DISCOVERY_SRV": runOpts.discoverySRV,
-	}, out, true); err != nil {
+	export := make(map[string]string)
+	etcdName := os.Getenv("ETCD_NAME")
+	if etcdName == "" {
+		return fmt.Errorf("environment variable ETCD_NAME has no value")
+	}
+	var e EtcdScaling
```
Comment (kikisdeliveryservice) [inline cmd/setup-etcd-environment/run.go]: could you use a more descriptive name than `e`, it's hard to remember what this stands for as i read thru below.

File: cmd/setup-etcd-environment/run.go (hunk 2 above)
Comment (hexfusion) [reply cmd/setup-etcd-environment/run.go]: This functionality is part of cluster bootstrap and relies on the operator so networking would be assumed. DR would require standing up a single node etcd cluster then scaling additional members via operator.

File: cmd/setup-etcd-environment/run.go (hunk 3 above)
Comment (hexfusion) [reply cmd/setup-etcd-environment/run.go]: I see we have glog will drop.

File: cmd/setup-etcd-environment/run.go (hunk 4 above)
Comment (hexfusion) [reply cmd/setup-etcd-environment/run.go]: good point will adjust

File: cmd/setup-etcd-environment/run.go (hunk 5 above)
Comment (hexfusion) [reply cmd/setup-etcd-environment/run.go]: sounds good.

File: cmd/setup-etcd-environment/run.go (hunk 2 above)
Comment (hexfusion) [reply cmd/setup-etcd-environment/run.go]: But I will consider how to handle this for that single node. thanks.

**Comment by hexfusion** [comment]
/close

Work is continued in #1221

**Pull request closed without merging**